### ISSUE TITLE
Extract Metadata

### ISSUE THREAD

**Issue by tcurdt** [hhrutter/pdfcpu#38]
I'd like to create a hash of PDF documents to compare them. The problem: PDFs use timestamps and UUIDs that might be different when you create the PDF - despite having the completely same structure and content. For more context check [here](https://github.com/jbarlow83/OCRmyPDF/issues/294)

I was wondering if I could parse the PDF and replace UUIDs and timestamps and combine/serialize all this into a digest I could then compare.

Do you think this would be possible with `pdfcpu`?

**Comment by hhrutter** [comment]
Anything‘s possible. Let me think about this and I will get back to you.

**Comment by hhrutter** [comment]
You are raising an interesting issue.

I still need to verify this but I am assuming that any PDF writer is free
to and possibly even mandatory has to set the proper time stamps in the
info dict. Pdfcpu is doing this other tools might do it the same way or
differently or not at all. Definitely something not to rely on.

I also need to get back to the pdf spec to check out the purpose and usage
of the ID which is indeed a UUID. If you are lucky it can just be a matter
of comparing the IDs of 2 PDFs to find out if they are equal.

Can you describe your use case?
I do not understand the motivation behind trying to compare an original pdf
with a processed version. Any processing of an input pdf that results in an
output pdf has to result in an output pdf that is different. I cannot think
of any reason why the output pdf should be the same as the input pdf. For
example if you optimize any pdf with pdfcpu it results into a pdf that’s
semantically the same but with different internals an optimized xreftable
and so forth so comparing these 2 does not make any sense. Their
fingerprint has to be different also because of the updated info dict.

I guess it all comes down to your very own  definition of equality between
2 PDFs. If you can lay that out there is a chance to implement this using
pdfcpu.

I can give you better feedback if you can share the details of your usecase.

**Comment by tcurdt** [comment]
I have PDF files coming in to archive them. These files are either scanned or generated. I now would like to prevent double archivals. For scanned documents that's probably out of scope because that's a much harder problem to solve (with some visual hashing or some AI thingy). But it would be great to at least have it for generated PDFs.

Now all of them are going through `ocrmypdf` that adds a text layer and converts them into PDF/A. The problem is now that the transformation is not idempotent as it does not use the original timestamps (what I believe should be fixed) and it re-creates UUIDs (which is probably much harder to fix). See the linked issue. It boils down to:

```
hash(a.pdf) == hash(a.pdf)
hash(transform(a.pdf)) != hash(transform(a.pdf))
```

One option would be to additionally store the hash of the incoming file - which I am not so keen on.

Another idea was: While I might not be able to make the transform idempotent I might be able to make the digest creation idempotent by filtering out time based data.

```
hash(a.pdf) == hash(a.pdf)
hash(transform(a.pdf)) != hash(transform(a.pdf))
hash(strip(transform(a.pdf))) == hash(strip(transform(a.pdf)))
```

They idea is not compare the original with the transformed PDF but rather two transformed PDFs. The digest should be the same if they were transformed from the same source.

I am not sure how feasible that actually is in reality - but I thought it's an interesting idea to explore.
I hope this makes a little more sense now?

**Comment by hhrutter** [comment]
I went back to the spec and there are no UUIDs involved in this but the PDF spec defines so called File Identifiers (14.4) for referencing external PDF files from a source PDF file. 

This ID entry in the trailer dict is an array of 2 strings representing MD5 checksums. The first checksum identifies the document as such and the second checksum has to be recalculated for any incremental change of the file.

The idea is for the original PDF creator to calculate an MD5 of the current timestamp, the file path, the file size and the values of the info dict entries and put this MD5 in both slots of the ID array. Any subsequent change to the file should reflect in a new calculated hash of position 2 in the ID array.
This allows any PDF file dealing with external fileIds to recognize file and find out whether it has been modified.

Now you could parse 2 PDFs, extract their ID array and draw your own conclusions:
* if the first array element matches for both files then you know that both files are versions of the same originator otherwise you are dealing with two completely different files.
* If on top one of the files has identical entries for both array elements this has to be the original.

The reasons I cannot recommend using this though for your usecase is:
* The ID array was built for the sole purpose for dealing with external file references from within a PDF. 
* The ID array is optional.
* All entries of the info dict are optional this includes any creation or modification timestamps.
 
This is from 14.4 of the spec:

> External file references may be made more reliable by including a file identifier (PDF 1.1) in the file and using it in addition to the normal platform-based file designation. Matching the identifier in the file reference with the one in the file confirms whether the correct file was found.

> File identifiers shall be defined by the optional ID entry in a PDF file’s trailer dictionary (see 7.5.5, “File Trailer”). The ID entry is optional but should be used. The value of this entry shall be an array of two byte strings. The first byte string shall be a permanent identifier based on the contents of the file at the time it was originally created and shall not change when the file is incrementally updated. The second byte string shall be a changing identifier based on the file’s contents at the time it was last updated. When a file is first written, both identifiers shall be set to the same value. If both identifiers match when a file reference is resolved, it is very likely that the correct and unchanged file has been found. If only the first identifier matches, a different version of the correct file has been found.

> To help ensure the uniqueness of file identifiers, they should be computed by means of a message digest algorithm such as MD5 (described in Internet RFC 1321, The MD5 Message-Digest Algorithm; see the Bibliography), using the following information:

> • The current time
• A string representation of the file’s location, usually a pathname
• The size of the file in bytes
• The values of all entries in the file’s document information dictionary

The lesson I learned out of this is that pdfcpu is not dealing correctly with the ID array for updates.

**Comment by tcurdt** [comment]
But shouldn't the file identifier be the same  in this scenario:
```
$ocrmypdf -l deu+eng in.pdf v1.pdf
$ocrmypdf -l deu+eng in.pdf v2.pdf
$ md5 v*.pdf 
MD5 (v1.pdf) = 346018ad8fdc6ae3e75bef99caeec694
MD5 (v2.pdf) = 9d6e08ba2e857e9b0228352cfc66bd18
```
When I compare the hex dumps I see a change in the modification date and what is even marked as uuid. Have a look at the diffs here https://github.com/jbarlow83/OCRmyPDF/issues/294

**Comment by hhrutter** [comment]
The UUIDs in the hexdump you are referring to belong to Metadata dict entries (basically relevant metadata written in XMP which is an XML derivate) which are optional in standard PDF, I cannot speak about PDF/A though. If there is a way for you to ensure metadata entries then you can rely on the contained metadata of course.

Like I said updating the info dict timestamps for creation and modification are totally under the discretion of the specific PDF Writer - they are optional and not to be relied upon.

As far as the `file ids` in your `ocrmypdf` example go I can only tell you that I don't know 😕.
I do not know this library and also do not know what kind of strategy it follows regarding timestamp modification in the info dict and regarding updating the ID array in the trailer dict. So you have to go back and look at the source code.

It is my stance that any PDF writer that writes back to the original file or creates a copy with or w/o modification has to maintain/update both the trailer ID array and all timestamps in the info dict. Although this is not a requirement of the spec (see above) I think it is common sense and `pdfcpu` will follow along those lines starting with the next release.

**Comment by tcurdt** [comment]
Sounds like it might really an issue how `ocrmypdf` handles the metadata then. So either I help fix this in `ocrmypdf` or use`pdfcpu` to strip the metadata for the hashing. Or plan b - keep the hash of the original file. I fear the later is the most simple and most foolproof.

Still I'd like to learn how to strip the metadata entries with `pdfcpu`.

I appreciate the help. Glad the discussion also let to something useful for `pdfcpu`

**Comment by hhrutter** [comment]
We can do this..
Thx 💚

**Comment by hhrutter** [comment]
fixed with v0.1.16
